### PR TITLE
MetaBackend

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
This PR implements a MetaBackend to load the qibocloud backend and list the available workers.

### CODE REVIEW

File: src/qibo_cloud_backends/__init__.py [hunk 1]
```diff
@@ -4,3 +4,45 @@
...
+                ValueError,
+                f"Unsupported worker, please use one among {WORKERS}.",
+            )
+
+    def list_available(self) -> dict:
+        """Lists all the available qibocloud backends."""
+        available_backends = {}
+        for worker in WORKERS:
+            try:
+                MetaBackend.load(worker=worker, token="")
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qibo_cloud_backends/__init__.py:43]: No, indeed, I think here the token is missing. I am pretty sure that `qiskit` performs a connection test when a `QickitClientBackend` is created. I don't know whether `qibo-client` supports a similar pinging mechanism, but that it is what we would need.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi the qibo-client should be working now, could you please check why tests are failing?

**Comment by BrunoLiegiBastonLiegi** [comment]
Mmmmh weird, I am still getting the `404` error with `qibo-client`:
```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://cloud.qibo.science/run_circuit/
```
even running the example in the readme gives me the `404`
``` python
import qibo
import qibo_client

# create the circuit you want to run
circuit = qibo.models.QFT(5)

# authenticate to server through the client instance
token = "xxxxxxx"
client = qibo_client.TII(token)

# run the circuit
result = client.run_circuit(circuit, nshots=1000, device="sim")
```

Furthermore, now the `QiskitClientBackend` is failing as well because apparently the `ibmq_qasm_simulator` is not available anymore as a backend on the `IBMQ` endpoint:
```
qiskit.providers.exceptions.QiskitBackendNotFoundError: 'No backend matches the criteria'
```
 maybe they changed the name, I am not sure, I will need to check.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi after updating the Qibo version, the issue with qibo-client is now fixed.
Could you please have another look at tests?

**Comment by BrunoLiegiBastonLiegi** [comment]
Yep, it works for me locally now. I  also changed the default platform for `ibmq` to `ibm_osaka` since the qasm simulator is no longer supported, it worked locally for me with my key, but the CI gets an `unauthorized login error`, can you check that the ibmq key in the secrets is updated?

**Comment by scarrazza** [comment]
Thanks @BrunoLiegiBastonLiegi. I have updated the token however it looks like there is a limit of 3 jobs per time:
https://github.com/qiboteam/qibo-cloud-backends/actions/runs/9710674360/job/26802944890?pr=7#step:11:1249
Therefore, I believe, we cannot run all CIs simulatenously...

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi could you please try to limit the qiskit job submission just to 1 CI setup, e.g. linux 3.11?

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi please have a look at my last commit. In principle it should pass, however ibm-q takes forever to complete the execution (approx. 3h), it would be much better to just simulate using their infrastructure.

**Comment by BrunoLiegiBastonLiegi** [comment]
Thanks @scarrazza, the last commits look ok to me, I am just wondering is there any particular reason why we perform the `qibo-client` test for `linux-3.11` only as well? I also noticed that you removed the connection test in the  `QiboClientBackend.__init__`, was it taking too long?
Regarding `IBMQ`, yeah I guess we could use their `local simulator` backend, but this way we do not test the connection which is probably one of the main thing you would like to test, being this a cloud backend.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi, indeed, the qibo-client can be executed in parallel, however I think qiskit is interfering with urllib3. For example, if I switch off qiskit tests, everything works well in parallel: https://github.com/qiboteam/qibo-cloud-backends/actions/runs/9716683003. However, if I simply enable the `test_set_backend`, I get a crash: https://github.com/qiboteam/qibo-cloud-backends/actions/runs/9716707407. Do you have any idea why?

**Comment by BrunoLiegiBastonLiegi** [comment]
Thanks @scarrazza I can see that everything is working fine now. Since the `ibmq` tests are slow and the resources we have there are limited, should we maybe limit in the workflows the time they are run? maybe not on every single push, it would be ideal to trigger them manually probably.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi indeed, we have applied some minor fixes. We can keep as it is, however before merging we should fix coverage and update the documentation on how to set backend.

**Comment by BrunoLiegiBastonLiegi** [comment]
I'll take care of that thanks

**Comment by alecandido** [comment]
@BrunoLiegiBastonLiegi you could even trigger them periodically, like once per month. It would ensure things are being tested, and at the rate you wish, saving you the trouble to do it manually (and keeping track of the last time you did it).

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule

You could run them manually on top of that, whenever you have a specific need.

**Comment by BrunoLiegiBastonLiegi** [comment]
Yeah I would do something like that, once a month seems reasonable. Then, I would remove the auto trigger, for tests at least, on push event, set a periodical one once a month and ideally each PR should run the tests at least once before being merged. Can you have a button in the PR to trigger the CI? I imagine the PR process to be something along the lines:

- implement new features
- test locally until everything works
- trigger manually the CI tests
- review
- fix what's there to fix 
- trigger manually the CI tests one last time before merging

Alternatively, we could decouple the `ibm-q` tests that we know to be time consuming and expensive and apply all of the above to those alone.

Regarding the environment variables for the tokens, instead, I would like to propose a change of naming:
```
"QIBO_CLIENT_TII_TOKEN" --> "QIBO_CLIENT_TOKEN"
"IBMQ_TOKEN" --> "QISKIT_CLIENT_TOKEN"
```
as in my opinion, they should be coupled to the client backend rather than the provider/platform.

**Comment by alecandido** [comment]
You can not have a PR button. Instead, what you can have is:
- PR-related events https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request (but no way to know "this is the final test before merging, but only if it passes", for GitHub it will be one commit as all the others)
- the usual button in the "Actions" tab, https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch, but you can manually pick the desired branch
- a label event, https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#label, or a PR comment, https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_comment-use-issue_comment (as it says, the name is `issue-comment`, just because PRs and issues are almost the same thing at some level)

that you can use to trigger the workflow.

A real button to be displayed in the web UI is not available, but I guess you can get really close with the other options (especially the label).

**Comment by scarrazza** [comment]
> ```
> "QIBO_CLIENT_TII_TOKEN" --> "QIBO_CLIENT_TOKEN"
> "IBMQ_TOKEN" --> "QISKIT_CLIENT_TOKEN"
> ```

I agree, in fact in the future we might have other labs linked to the qibo-client, therefore no need to create separate tokens.

**Comment by BrunoLiegiBastonLiegi** [comment]
@alecandido do you have any idea why /run_tests is not triggering the tests? Am I missing something?

**Comment by alecandido** [comment]
> @alecandido do you have any idea why /run_tests is not triggering the tests? Am I missing something?

The way I read this
https://github.com/qiboteam/qibo-cloud-backends/blob/b4c5adfca68a911063b0e741383f47b25ac1b5b2/.github/workflows/unit_test.yml#L17
is: "run this job if the label contains `run-workflow` or if it is a `push` event".

In principle, your event has no label, and it should be an `issue_comment` event. So, the workflow should not get triggered.

Given that the label was unused, maybe you want to replace the condition on the label with the one on the comment (or just add the one about the comment).

**Comment by scarrazza** [comment]
For the sake of symmetry with other repositories, may I suggest to use the label mechanism instead of comment?

File: .github/workflows/unit_test.yml [hunk 2]
```diff
@@ -55,12 +53,11 @@ jobs:
           poe lint
           poe lint-warnings
       - name: Test
-        if: contains(github.event.comment.body, '/run_tests')  # check if the comment calls the tests execution
```
Comment (alecandido) [inline .github/workflows/unit_test.yml]: But now you should replace this with the label query (and schedule, whenever there will be a schedule event).

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi I would like to merge this PR asap and release.
Could you please double check if everything is in place?

**Comment by BrunoLiegiBastonLiegi** [comment]
Yeah sorry I am traveling for a conference this week and my time is going to be limited. Anyway, the PR looks mostly ready to me: the tests now trigger only when the run-workflow label is added, what is missing is the periodical test in case we want to add it and change of the environment variables name.

**Comment by alecandido** [comment]
Since there is no trigger on `push:` any longer, I agree with @BrunoLiegiBastonLiegi.

Eventually, I'd split some tests that could be executed on `push:` (even mocking) from the remote ones (to be executed with the current events). And add the scheduling.
But all of this could be done in separate PRs, so I'd just merge this one (which was targeted to support the `MetaBackend` mechanism, not to improve the testing one).

**Review by alecandido — APPROVED** [review]

**Comment by scarrazza** [comment]
Thanks @BrunoLiegiBastonLiegi and @alecandido. I will merge and release this repo now, please open the respective issues concerning the workflows.

**Pull request merged**